File: DP/match_coin.cpp
#include <iostream>
#include <vector>

using namespace std;

int main(void) {
    int n, price;
    cin >> n >> price;

    vector<int> coins;

    for (int i = 0; i < n; i++) {
        int x;
        cin >> x;
        coins.push_back(x);
    }

    vector<int> dp(price + 1, price); // DP 초기화
    // cpp의 벡터 초기화 (크기, 값)

    for (int i = 1; i < price + 1; ++i) {
        for (int j = 0; j < n; ++j) {
            if (i == coins[j]) {
                dp[i] = 1;
            } else if (i < coins[j]) {
                dp[i] = -1;
            } else if (i > coins[j]) {
                int remainder = i - coins[j];
                if (dp[remainder] != -1) {
                    dp[i] = min(dp[i], dp[remainder] + 1);
                } else {
                    dp[i] = -1;
                }
            }
        }
    }
}

File: DP/match_coin.py
import sys

rl = lambda: sys.stdin.readline().rstrip()

n, price = map(int, rl().split())

coins = []
for _ in range(n):
    coins.append(int(rl()))

dp = [price] * (price + 1)

for i in range(1, price + 1):
    for coin in coins:
        if i == coin:
            dp[i] = 1
        elif i < coin:
            dp[i] = -1
        elif i > coin:
            remainder = i - coin
            if dp[remainder] != -1:
                dp[i] = min(dp[i], dp[remainder] + 1)
            else: dp[i] = -1

print(dp[price])
